content/academy/puppeteer_playwright/page/waiting.md

```diff
@@ -0,0 +1,136 @@
+
+```JavaScript
+import puppeteer from 'puppeteer';
+
+const browser = await puppeteer.launch({ headless: false });
+
+const page = await browser.newPage();
+await page.goto('https://google.com/');
+
+await page.click('button + button:nth-child(2)');
+
+await page.type('input[title="Search"]', 'hello world');
+await page.keyboard.press('Enter');
+
+// Wait for the element to be present on the page prior to clicking it
+await page.waitForSelector('.g a');
+await page.click('.g a');
+
+await page.waitForTimeout(10000)
+await browser.close();
+```
+
+Now, we won't see the error message anymore, and the first result will be successfully clicked by Puppeteer.
+
+## [](#waiting-for-navigation) Navigation
+
+If we remember properly, after clicking the first result, we want to console log the title of the result's page and save a screenshot into the filesystem. In order to grab a solid screenshot of the loaded page though, we should **wait for navigation**  before snapping the image. This can be done with `page.waitForNavigation()`.
+
+Naively, you might immediately think that this is the way we should wait for navigation after clicking the first result:
+
+```JavaScript
+await page.click('.g a');
+await page.waitForNavigation();
+```
+
+Though in theory this is correct, it can result in a race condition in which the page navigates quickly before the `page.waitForNavigation()` function is ever run, which means that once it is finally called, it will hang and wait forever for the `domcontentloaded` event to fire even though it already fired. To solve this, we can stick the waiting logic and the clicking logic into a `Promise.all()` call (placing `page.waitForNavigation()` first).
+
+```JavaScript
+await Promise.all([page.waitForNavigation(), page.click('.g a')]);
+```
+
+This implementation will do the following:
+
+1. Begin waiting for the page to navigate without blocking the `page.click()` function
+2. Click the element, firing off a navigating event
+3. Resolve once the page has navigated, allowing further code to run
+
+## [](#current-code) Our code so far
+
+Here's what our project's code looks like so far:
+
+```marked-tabs
+<marked-tab header="Playwright" lang="javascript">
+import puppeteer from 'puppeteer';
+import * as fs from 'fs/promises';
+
+const browser = await puppeteer.launch({ headless: false });
+
+// Create a page and visit Google
+const page = await browser.newPage();
+await page.goto('https://google.com');
+
+// Agree to the cookies policy
+await page.click('button + button:nth-child(2)');
+
+// Type the query and visit the results page
+await page.type('input[title="Search"]', 'hello world');
+await page.keyboard.press('Enter');
+
+// Wait for the first result to appear on the page,
+// then click on it
+await page.waitForSelector('.g a');
+await Promise.all([page.waitForNavigation(), page.click('.g a')]);
```

Comment:
Would replace with waitForLoadState. Yeah, I know, sorry that I didn't mention this earlier, but I didn't know about this function before. But it's better, so let's use it.